File: tests/test_core.py
import json
import os
import subprocess
import sys
import tempfile
import unittest
from pathlib import Path
from unittest.mock import patch

ROOT = Path(__file__).resolve().parents[1]
if str(ROOT) not in sys.path:
    sys.path.insert(0, str(ROOT))

from template_to_pdf.builder import build, fill
from template_to_pdf.core import render
from template_to_pdf.extract import PROMPT, _parse_json, extract
from template_to_pdf.schema import flatten, normalize


class TestSchema(unittest.TestCase):
    def test_normalize_valid(self):
        doc = normalize({
            "title": "Report",
            "sections": [{"heading": "Intro", "body": "Hello"}],
            "meta": {"author": "Alice", "version": 1},
        })
        self.assertEqual(doc["title"], "Report")
        self.assertEqual(len(doc["sections"]), 1)
        self.assertEqual(doc["sections"][0], {"heading": "Intro", "body": "Hello"})
        self.assertEqual(doc["meta"]["author"], "Alice")
        self.assertEqual(doc["meta"]["version"], "1")

    def test_normalize_defaults(self):
        doc = normalize({})
        self.assertEqual(doc["title"], "")
        self.assertEqual(doc["sections"], [])
        self.assertEqual(doc["meta"], {})

    def test_normalize_none_fields(self):
        doc = normalize({"title": None, "sections": None, "meta": None})
        self.assertEqual(doc["title"], "")
        self.assertEqual(doc["sections"], [])
        self.assertEqual(doc["meta"], {})

    def test_normalize_coercion(self):
        doc = normalize({
            "title": 12345,
            "sections": [{"heading": 42, "body": True}, {"heading": None, "body": None}],
            "meta": {100: False},
        })
        self.assertEqual(doc["title"], "12345")
        self.assertEqual(doc["sections"][0], {"heading": "42", "body": "True"})
        self.assertEqual(doc["sections"][1], {"heading": "", "body": ""})
        self.assertEqual(doc["meta"]["100"], "False")

    def test_normalize_invalid_root_type(self):
        for bad in [None, 123, "string", [1, 2, 3]]:
            with self.subTest(bad=bad):
                with self.assertRaises(TypeError):
                    normalize(bad)

    def test_normalize_invalid_sections(self):
        with self.assertRaises(TypeError):
            normalize({"sections": "not-a-list"})
        with self.assertRaises(TypeError):
            normalize({"sections": ["not-a-dict"]})

    def test_normalize_invalid_meta(self):
        with self.assertRaises(TypeError):
            normalize({"meta": "not-a-dict"})
        with self.assertRaises(TypeError):
            normalize({"meta": [1, 2]})

    def test_flatten_happy_path(self):
        doc = {
            "title": "Title",
            "sections": [
                {"heading": "H1", "body": "B1"},
                {"heading": "H2", "body": "B2"},
            ],
            "meta": {"author": "Bob", "date": "2026-09-10"},
        }
        vars_ = flatten(doc)
        self.assertEqual(vars_["title"], "Title")
        self.assertIn("<h2>H1</h2>\n<p>B1</p>", vars_["body"])
        self.assertIn("<h2>H2</h2>\n<p>B2</p>", vars_["body"])
        self.assertEqual(vars_["author"], "Bob")
        self.assertEqual(vars_["date"], "2026-09-10")

    def test_flatten_no_sections(self):
        vars_ = flatten({"title": "Empty"})
        self.assertEqual(vars_["title"], "Empty")
        self.assertEqual(vars_["body"], "")

    def test_flatten_conflict_protection(self):
        vars_ = flatten({
            "title": "Original Title",
            "sections": [{"heading": "H", "body": "B"}],
            "meta": {"title": "Overwritten Title", "body": "Overwritten Body", "custom": "Keep"},
        })
        self.assertEqual(vars_["title"], "Original Title")
        self.assertIn("<h2>H</h2>\n<p>B</p>", vars_["body"])
        self.assertEqual(vars_["custom"], "Keep")

    def test_unicode_and_special_chars(self):
        data = {
            "title": "报告 Title 🎉",
            "sections": [{"heading": "第 1 节", "body": "内容 with <tags> & 'quotes'"}],
            "meta": {"emoji": "🚀"},
        }
        doc = normalize(data)
        vars_ = flatten(doc)
        self.assertEqual(vars_["title"], "报告 Title 🎉")
        self.assertIn("第 1 节", vars_["body"])
        self.assertEqual(vars_["emoji"], "🚀")


class TestExtract(unittest.TestCase):
    def test_parse_json_direct(self):
        val = _parse_json('{"title": "Clean"}')
        self.assertEqual(val, {"title": "Clean"})

    def test_parse_json_fenced(self):
        val = _parse_json('```json\n{"title": "Fenced"}\n```')
        self.assertEqual(val, {"title": "Fenced"})

    def test_parse_json_fenced_untagged(self):
        val = _parse_json('```\n{"title": "Untagged"}\n```')
        self.assertEqual(val, {"title": "Untagged"})

    def test_parse_json_multiple_fences(self):
        text = "Here is an example:\n```python\nprint(1)\n```\nHere is JSON:\n```json\n{\"title\": \"Second\"}\n```"
        val = _parse_json(text)
        self.assertEqual(val, {"title": "Second"})

    def test_parse_json_invalid(self):
        self.assertIsNone(_parse_json(None))
        self.assertIsNone(_parse_json(12345))
        self.assertIsNone(_parse_json(""))
        self.assertIsNone(_parse_json("   "))
        self.assertIsNone(_parse_json("not valid json"))
        self.assertIsNone(_parse_json("[1, 2, 3]"))
        self.assertIsNone(_parse_json('"just a string"'))

    def test_extract_direct_json(self):
        doc = extract('{"title": "T", "sections": []}')
        self.assertEqual(doc["title"], "T")

    def test_extract_no_llm_fails_on_plain_text(self):
        with self.assertRaises(ValueError):
            extract("raw text without json")

    def test_extract_sync_llm(self):
        called_prompt = []

        def mock_llm(p):
            called_prompt.append(p)
            return '{"title": "From LLM", "sections": [{"heading": "A", "body": "B"}]}'

        doc = extract("my raw notes", llm=mock_llm)
        self.assertEqual(doc["title"], "From LLM")
        self.assertEqual(doc["sections"][0]["heading"], "A")
        self.assertTrue(called_prompt[0].startswith(PROMPT))
        self.assertIn("my raw notes", called_prompt[0])

    def test_extract_llm_returns_fenced(self):
        doc = extract("notes", llm=lambda p: '```json\n{"title": "Fenced LLM"}\n```')
        self.assertEqual(doc["title"], "Fenced LLM")

    def test_extract_llm_invalid_json(self):
        with self.assertRaises(ValueError):
            extract("notes", llm=lambda p: "sorry I cannot generate JSON")


class TestBuilder(unittest.TestCase):
    def test_fill_variables(self):
        tmpl = "Hello {{name}}, welcome to {{ place }}! Missing: {{unknown}}"
        res = fill(tmpl, {"name": "World", "place": "Earth"})
        self.assertEqual(res, "Hello World, welcome to Earth! Missing: ")

    def test_fill_single_pass(self):
        tmpl = "{{var}}"
        res = fill(tmpl, {"var": "{{recursive}}"})
        self.assertEqual(res, "{{recursive}}")

    def test_build_html_full_template(self):
        with tempfile.NamedTemporaryFile("w", suffix=".html", delete=False, encoding="utf-8") as f:
            f.write("<!doctype html><html><head><title>{{title}}</title></head><body><h1>{{title}}</h1>{{body}}</body></html>")
            path = f.name
        try:
            html = build(path, {"title": "Doc", "sections": [{"heading": "H", "body": "B"}]})
            self.assertEqual(html.count("<!doctype html>"), 1)
            self.assertIn("<h1>Doc</h1>", html)
            self.assertIn("<h2>H</h2>", html)
        finally:
            os.remove(path)

    def test_build_html_fragment(self):
        with tempfile.NamedTemporaryFile("w", suffix=".html", delete=False, encoding="utf-8") as f:
            f.write("<h1>{{title}}</h1><div>{{body}}</div>")
            path = f.name
        try:
            html = build(path, {"title": "Fragment", "sections": []})
            self.assertIn("<!doctype html>", html.lower())
            self.assertIn("<title>Fragment</title>", html)
            self.assertIn("<h1>Fragment</h1>", html)
        finally:
            os.remove(path)

    def test_build_markdown_template(self):
        with tempfile.NamedTemporaryFile("w", suffix=".md", delete=False, encoding="utf-8") as f:
            f.write("# {{title}}\n\n**Bold note**\n\n{{body}}")
            path = f.name
        try:
            html = build(path, {"title": "MD Doc", "sections": [{"heading": "Sec1", "body": "Text1"}]})
            self.assertIn("<!doctype html>", html.lower())
            self.assertIn("<h1>MD Doc</h1>", html)
            self.assertIn("<strong>Bold note</strong>", html)
            self.assertIn("<h2>Sec1</h2>", html)
        finally:
            os.remove(path)

    def test_build_missing_template(self):
        with self.assertRaises(FileNotFoundError):
            build("non_existent_template.html", {"title": "T"})


class TestCore(unittest.TestCase):
    @patch("template_to_pdf.core.to_pdf")
    def test_render_pipeline(self, mock_to_pdf):
        with tempfile.TemporaryDirectory() as tmpdir:
            tmpl_path = Path(tmpdir) / "test.html"
            tmpl_path.write_text("<h1>{{title}}</h1>{{body}}", encoding="utf-8")
            pdf_path = Path(tmpdir) / "output.pdf"

            out = render('{"title": "Core", "sections": [{"heading": "H", "body": "B"}]}', str(tmpl_path), str(pdf_path))
            self.assertEqual(out, str(pdf_path))
            mock_to_pdf.assert_called_once()
            called_html, called_path = mock_to_pdf.call_args[0]
            self.assertIn("<h1>Core</h1>", called_html)
            self.assertEqual(called_path, str(pdf_path))


class TestCLI(unittest.TestCase):
    def test_cli_success(self):
        with tempfile.TemporaryDirectory() as tmpdir:
            input_file = Path(tmpdir) / "input.json"
            input_file.write_text(json.dumps({"title": "CLI Test", "sections": []}), encoding="utf-8")
            tmpl_file = Path(tmpdir) / "tmpl.html"
            tmpl_file.write_text("<h1>{{title}}</h1>", encoding="utf-8")
            out_file = Path(tmpdir) / "out.pdf"

            with patch("template_to_pdf.core.to_pdf") as mock_to_pdf:
                from template_to_pdf.__main__ import main
                with patch.object(sys, "argv", ["template-to-pdf", "--input", str(input_file), "--template", str(tmpl_file), "--out", str(out_file)]):
                    main()
                mock_to_pdf.assert_called_once()

    def test_cli_missing_args(self):
        with patch.object(sys, "argv", ["template-to-pdf"]):
            with self.assertRaises(SystemExit) as cm:
                from template_to_pdf.__main__ import main
                main()
            self.assertNotEqual(cm.exception.code, 0)


if __name__ == "__main__":
    unittest.main()

File: tests/test_core.js
import test from "node:test"
import assert from "node:assert/strict"
import { mkdtempSync, rmSync, writeFileSync, existsSync } from "node:fs"
import { tmpdir } from "node:os"
import { join, dirname } from "node:path"
import { fileURLToPath } from "node:url"
import { execFile } from "node:child_process"
import { promisify } from "node:util"

import { normalize, flatten } from "../js/schema.js"
import { extract, parseJson, PROMPT } from "../js/extract.js"
import { fill, build } from "../js/builder.js"
import { render } from "../js/core.js"
import { toPdf } from "../js/converter.js"

const execFileAsync = promisify(execFile)
const ROOT = join(dirname(fileURLToPath(import.meta.url)), "..")

test("Schema Suite", async (t) => {
  await t.test("normalize valid document", () => {
    const doc = normalize({
      title: "Report",
      sections: [{ heading: "Intro", body: "Hello" }],
      meta: { author: "Alice", version: 1 },
    })
    assert.equal(doc.title, "Report")
    assert.equal(doc.sections.length, 1)
    assert.deepEqual(doc.sections[0], { heading: "Intro", body: "Hello" })
    assert.equal(doc.meta.author, "Alice")
    assert.equal(doc.meta.version, "1")
  })

  await t.test("normalize defaults on empty object", () => {
    const doc = normalize({})
    assert.equal(doc.title, "")
    assert.deepEqual(doc.sections, [])
    assert.deepEqual(doc.meta, {})
  })

  await t.test("normalize null fields safely", () => {
    const doc = normalize({ title: null, sections: null, meta: null })
    assert.equal(doc.title, "")
    assert.deepEqual(doc.sections, [])
    assert.deepEqual(doc.meta, {})
  })

  await t.test("normalize type coercion", () => {
    const doc = normalize({
      title: 12345,
      sections: [{ heading: 42, body: true }, { heading: null, body: null }],
      meta: { 100: false },
    })
    assert.equal(doc.title, "12345")
    assert.equal(doc.sections[0].heading, "42")
    assert.equal(doc.sections[0].body, "true")
    assert.equal(doc.sections[1].heading, "")
    assert.equal(doc.sections[1].body, "")
    assert.equal(doc.meta["100"], "false")
  })

  await t.test("normalize rejects non-object root", () => {
    for (const bad of [null, undefined, 123, "string", [1, 2]]) {
      assert.throws(() => normalize(bad), { name: "TypeError", message: "expected object" })
    }
  })

  await t.test("normalize rejects invalid sections", () => {
    assert.throws(() => normalize({ sections: "not-a-list" }), { name: "TypeError" })
    assert.throws(() => normalize({ sections: ["not-an-object"] }), { name: "TypeError" })
  })

  await t.test("normalize rejects invalid meta", () => {
    assert.throws(() => normalize({ meta: "not-an-object" }), { name: "TypeError" })
    assert.throws(() => normalize({ meta: [1, 2] }), { name: "TypeError" })
  })

  await t.test("flatten sections and meta", () => {
    const vars = flatten({
      title: "Title",
      sections: [
        { heading: "H1", body: "B1" },
        { heading: "H2", body: "B2" },
      ],
      meta: { author: "Bob", date: "2026-09-10" },
    })
    assert.equal(vars.title, "Title")
    assert.match(vars.body, /<h2>H1<\/h2>\n<p>B1<\/p>/)
    assert.match(vars.body, /<h2>H2<\/h2>\n<p>B2<\/p>/)
    assert.equal(vars.author, "Bob")
    assert.equal(vars.date, "2026-09-10")
  })

  await t.test("flatten with zero sections", () => {
    const vars = flatten({ title: "Empty" })
    assert.equal(vars.title, "Empty")
    assert.equal(vars.body, "")
  })

  await t.test("flatten conflict protection", () => {
    const vars = flatten({
      title: "Original Title",
      sections: [{ heading: "H", body: "B" }],
      meta: { title: "Overwritten Title", body: "Overwritten Body", custom: "Keep" },
    })
    assert.equal(vars.title, "Original Title")
    assert.match(vars.body, /<h2>H<\/h2>\n<p>B<\/p>/)
    assert.equal(vars.custom, "Keep")
  })

  await t.test("unicode and special characters", () => {
    const doc = normalize({
      title: "报告 Title 🎉",
      sections: [{ heading: "第 1 节", body: "内容 with <tags> & 'quotes'" }],
      meta: { emoji: "🚀" },
    })
    const vars = flatten(doc)
    assert.equal(vars.title, "报告 Title 🎉")
    assert.match(vars.body, /第 1 节/)
    assert.equal(vars.emoji, "🚀")
  })
})

test("Extract Suite", async (t) => {
  await t.test("parseJson direct string", () => {
    const val = parseJson('{"title": "Clean"}')
    assert.deepEqual(val, { title: "Clean" })
  })

  await t.test("parseJson fenced markdown", () => {
    const val = parseJson('```json\n{"title": "Fenced"}\n```')
    assert.deepEqual(val, { title: "Fenced" })
  })

  await t.test("parseJson fenced untagged", () => {
    const val = parseJson('```\n{"title": "Untagged"}\n```')
    assert.deepEqual(val, { title: "Untagged" })
  })

  await t.test("parseJson multiple fences picks valid json", () => {
    const text = "Note:\n```python\nprint(1)\n```\nResult:\n```json\n{\"title\": \"Second\"}\n```"
    const val = parseJson(text)
    assert.deepEqual(val, { title: "Second" })
  })

  await t.test("parseJson invalid inputs return null", () => {
    assert.equal(parseJson(null), null)
    assert.equal(parseJson(12345), null)
    assert.equal(parseJson(""), null)
    assert.equal(parseJson("   "), null)
    assert.equal(parseJson("not json"), null)
    assert.equal(parseJson("[1, 2, 3]"), null)
    assert.equal(parseJson('"string only"'), null)
  })

  await t.test("extract direct JSON without LLM", () => {
    const doc = extract('{"title": "Direct", "sections": []}')
    assert.equal(doc.title, "Direct")
  })

  await t.test("extract plain text without LLM throws", () => {
    assert.throws(() => extract("plain notes without json"), { message: /non-JSON/ })
  })

  await t.test("extract with sync LLM", () => {
    let calledPrompt = ""
    const doc = extract("raw text", (p) => {
      calledPrompt = p
      return '{"title": "Sync LLM", "sections": [{"heading": "A", "body": "B"}]}'
    })
    assert.equal(doc.title, "Sync LLM")
    assert.equal(doc.sections[0].heading, "A")
    assert.ok(calledPrompt.startsWith(PROMPT))
    assert.ok(calledPrompt.includes("raw text"))
  })

  await t.test("extract with async LLM", async () => {
    const promise = extract("raw text", async (p) => {
      await new Promise((r) => setTimeout(r, 10))
      return '```json\n{"title": "Async LLM", "sections": []}\n```'
    })
    assert.ok(promise instanceof Promise)
    const doc = await promise
    assert.equal(doc.title, "Async LLM")
  })

  await t.test("extract with LLM returning invalid JSON throws", () => {
    assert.throws(() => extract("raw text", () => "invalid response"), { message: /did not return JSON/ })
  })
})

test("Builder Suite", async (t) => {
  await t.test("fill variables", () => {
    const tmpl = "Hello {{name}}, welcome to {{ place }}! Missing: {{unknown}}"
    const res = fill(tmpl, { name: "World", place: "Earth" })
    assert.equal(res, "Hello World, welcome to Earth! Missing: ")
  })

  await t.test("fill is single-pass", () => {
    const tmpl = "{{var}}"
    const res = fill(tmpl, { var: "{{recursive}}" })
    assert.equal(res, "{{recursive}}")
  })

  await t.test("build HTML full document", () => {
    const tmp = mkdtempSync(join(tmpdir(), "t2p-"))
    const tmplPath = join(tmp, "full.html")
    writeFileSync(tmplPath, "<!doctype html><html><head><title>{{title}}</title></head><body><h1>{{title}}</h1>{{body}}</body></html>")
    try {
      const html = build(tmplPath, { title: "Full", sections: [{ heading: "H", body: "B" }] })
      assert.equal((html.match(/<!doctype html>/gi) || []).length, 1)
      assert.match(html, /<h1>Full<\/h1>/)
      assert.match(html, /<h2>H<\/h2>/)
    } finally {
      rmSync(tmp, { recursive: true, force: true })
    }
  })

  await t.test("build HTML fragment wraps in html5", () => {
    const tmp = mkdtempSync(join(tmpdir(), "t2p-"))
    const tmplPath = join(tmp, "frag.html")
    writeFileSync(tmplPath, "<h1>{{title}}</h1><div>{{body}}</div>")
    try {
      const html = build(tmplPath, { title: "Frag", sections: [] })
      assert.match(html.toLowerCase(), /<!doctype html>/)
      assert.match(html, /<title>Frag<\/title>/)
      assert.match(html, /<h1>Frag<\/h1>/)
    } finally {
      rmSync(tmp, { recursive: true, force: true })
    }
  })

  await t.test("build Markdown converts and wraps in html5", () => {
    const tmp = mkdtempSync(join(tmpdir(), "t2p-"))
    const tmplPath = join(tmp, "doc.md")
    writeFileSync(tmplPath, "# {{title}}\n\n**Bold note**\n\n{{body}}")
    try {
      const html = build(tmplPath, { title: "MD Doc", sections: [{ heading: "Sec1", body: "Text1" }] })
      assert.match(html.toLowerCase(), /<!doctype html>/)
      assert.match(html, /<h1>MD Doc<\/h1>/)
      assert.match(html, /<strong>Bold note<\/strong>/)
      assert.match(html, /<h2>Sec1<\/h2>/)
    } finally {
      rmSync(tmp, { recursive: true, force: true })
    }
  })

  await t.test("build non-existent template throws", () => {
    assert.throws(() => build("non_existent.html", { title: "T" }), { code: "ENOENT" })
  })
})

test("Core & CLI Suite", async (t) => {
  await t.test("render pipeline with async LLM and mocked PDF", async () => {
    const tmp = mkdtempSync(join(tmpdir(), "t2p-"))
    const tmplPath = join(tmp, "t.html")
    const pdfPath = join(tmp, "out.pdf")
    writeFileSync(tmplPath, "<h1>{{title}}</h1>{{body}}")

    try {
      const out = await render(
        "raw text",
        tmplPath,
        pdfPath,
        async () => '{"title": "Core Render", "sections": [{"heading": "H", "body": "B"}]}'
      )
      assert.equal(out, pdfPath)
      assert.ok(existsSync(pdfPath))
    } finally {
      rmSync(tmp, { recursive: true, force: true })
    }
  })

  await t.test("CLI execution success", async () => {
    const tmp = mkdtempSync(join(tmpdir(), "t2p-"))
    const inputPath = join(tmp, "in.json")
    const tmplPath = join(tmp, "t.html")
    const pdfPath = join(tmp, "cli-out.pdf")

    writeFileSync(inputPath, JSON.stringify({ title: "CLI", sections: [] }))
    writeFileSync(tmplPath, "<h1>{{title}}</h1>")

    try {
      const { stdout } = await execFileAsync("node", [
        join(ROOT, "js", "cli.js"),
        "--input",
        inputPath,
        "--template",
        tmplPath,
        "--out",
        pdfPath,
      ])
      assert.ok(existsSync(pdfPath))
    } finally {
      rmSync(tmp, { recursive: true, force: true })
    }
  })

  await t.test("CLI missing arguments fails", async () => {
    await assert.rejects(
      async () => {
        await execFileAsync("node", [join(ROOT, "js", "cli.js")])
      },
      (err) => {
        assert.equal(err.code, 1)
        assert.match(err.stderr, /usage: template-to-pdf/)
        return true
      }
    )
  })
})
